plutus-core/testlib/PlutusIR/Properties/Evaluation.hs

```diff
@@ -0,0 +1,150 @@
+import PlutusCore.Default
+import PlutusCore.Evaluation.Machine.Ck
+import PlutusCore.Evaluation.Machine.ExBudgetingDefaults
+import PlutusCore.Generators.QuickCheck.Builtin
+import PlutusCore.Generators.QuickCheck.GenTm
+import PlutusCore.Generators.QuickCheck.Utils
+import PlutusCore.Pretty
+import PlutusCore.Quote
+import PlutusCore.Rename
+import PlutusCore.Rename.Monad
+import PlutusCore.TypeCheck qualified as PLC
+import PlutusIR.Compiler
+import PlutusIR.Compiler.Lower
+import PlutusIR.Core
+import PlutusIR.Generators.QuickCheck.GenerateTerms
+import PlutusPrelude (($>))
+import Test.QuickCheck (Property)
+
+-- Convert the Either from the evaluation to Either String () to match with the
+-- Testable (Either String ()) instance.
+convertToEitherString ::
+  Either
+    (CkEvaluationException DefaultUni DefaultFun) (PLC.Term TyName Name DefaultUni DefaultFun ())
+  -> Either String (PLC.Term TyName Name DefaultUni DefaultFun ())
+convertToEitherString = \case
+  Left _   -> Left $ show "eqTermM returns Nothing"
+  Right tm -> Right tm
+
+-- | Evaluate a PIR term after a pure compiler pass.
+evalPurePass ::
+  (Term TyName Name DefaultUni DefaultFun (Provenance ())
+  -> Term TyName Name DefaultUni DefaultFun (Provenance ()))
+  -> BuiltinSemanticsVariant DefaultFun
+  -> Term TyName Name DefaultUni DefaultFun (Provenance ()) ->
+  Either
+    String
+    (PLC.Term TyName Name DefaultUni DefaultFun ())
+evalPurePass pass biVariant pirTm = do
+  -- let plcConfig = first display $ PLC.getDefTypeCheckConfig ()
+  plcConfig <- PLC.getDefTypeCheckConfig (Original ())
+  lowered <- flip runReaderT (toDefaultCompilationCtx plcConfig) $
+    runQuoteT $ lowerTerm $ pass (Original () <$ pirTm)
+  convertToEitherString $
+    evaluateCkNoEmit (defaultBuiltinsRuntimeForSemanticsVariant biVariant) (lowered $> ())
+
+-- | Evaluate a PIR term that is lowered to a PLC term without any PIR pass.
+evalNoPass :: BuiltinSemanticsVariant DefaultFun
+  -> Term TyName Name DefaultUni DefaultFun (Provenance ()) ->
+  Either String (PLC.Term TyName Name DefaultUni DefaultFun ())
+evalNoPass = evalPurePass id
+
+-- | Evaluate a PIR term after a pure compiler pass.
+evalNonPurePass ::
+  (Term TyName Name DefaultUni DefaultFun (Provenance ())
+  -> QuoteT
+      (ReaderT (CompilationCtx DefaultUni DefaultFun ()) (Either String))
+      (Term TyName Name DefaultUni DefaultFun (Provenance ())))
+  -> BuiltinSemanticsVariant DefaultFun
+  -> Term TyName Name DefaultUni DefaultFun b
+  -> Either String (PLC.Term TyName Name DefaultUni DefaultFun ())
+evalNonPurePass pass biVariant pirTm = do
+  plcConfig <- PLC.getDefTypeCheckConfig (Original ())
+  lowered <- flip runReaderT (toDefaultCompilationCtx plcConfig) $
+    runQuoteT $ lowerTerm =<< pass (Original () <$ pirTm)
+  convertToEitherString $
+    evaluateCkNoEmit (defaultBuiltinsRuntimeForSemanticsVariant biVariant) (lowered $> ())
+
+-- | Generate builtin types. See this module's haddock.
```

Comment:
This is how I generate the terms with `TyBuiltin` - but the generated terms included `Let` terms, as seen from the error from the deadcode test that I added. What am I doing wrong?

Reply:
The error:
```
 Evaluation:                                                                   FAIL
          *** Failed! Error during compilation: Let bindings should have been eliminated before lowering(()) (after 2 tests):
          DefaultFunSemanticsVariant1
          DefaultFunSemanticsVariant2
          ty,tm = let
            ~f2_2 : data = B #
          in
          1
```